mimesis/providers/base.py

```diff
@@ -74,9 +79,83 @@ def __init__(self, locale: Optional[str] = None,
         :param seed: Seed to all the random functions.
         """
         super().__init__(seed=seed)
-        self._data: dict
-        self._datafile: str
-        self.locale = setup_locale(locale)
+        self._data: dict = {}
+        self._datafile: Optional[str] = None
+        self._setup_locale(locale)
+        # TODO: Use pathlib instead
+        self._data_dir: Union[str, Path] = os.path.abspath(
+            os.path.join(os.path.join(
+                os.path.dirname(__file__), os.pardir), 'data'))
+
+    def _setup_locale(self, locale: Optional[str] = None) -> None:
+        """Set up locale after pre-check.
+
+        :param str locale: Locale
+        :raises UnsupportedLocale: When locale is not supported.
+        :return: Nothing.
+        """
+        if not locale:
+            locale = locales.DEFAULT_LOCALE
+
+        locale = locale.lower()
+        if locale not in locales.SUPPORTED_LOCALES:
+            raise UnsupportedLocale(locale)
+
+        self.locale = locale
+
+    def _update_dict(self, initial: JSON, other: Mapping) -> JSON:
+        """Recursively update a dictionary.
+
+        :param initial: Dict to update.
+        :param other: Dict to update from.
+        :return: Updated dict.
+        """
+        for key, value in other.items():
+            if isinstance(value, collections.Mapping):
+                r = self._update_dict(initial.get(key, {}), value)
+                initial[key] = r
+            else:
+                initial[key] = other[key]
+        return initial
+
+    @functools.lru_cache(maxsize=None)
+    def pull(self, datafile: Optional[str] = None):
+        """Pull the content from the JSON and memorize one.
+
+        Opens JSON file ``file`` in the folder ``data/locale``
+        and get content from the file and memorize ones using lru_cache.
+
+        :param datafile: The name of file.
+        :return: The content of the file.
+        :raises UnsupportedLocale: if locale is not supported.
+        """
+        locale = self.locale
+        data_dir = self._data_dir
+
+        if not datafile:
+            datafile = self._datafile
+
+        def get_data(locale_name: str) -> JSON:
+            """Pull JSON data from file.
+
+            :param locale_name: Locale name.
+            :return: Content of JSON file as dict.
+            """
+            file_path = os.path.join(data_dir, locale_name, datafile)
+            with open(file_path, 'r', encoding='utf8') as f:
```

Comment:
where checking if file exist?

Reply:
@ivlevdenis, `get_data` function called after:
```python
        if locale not in locales.SUPPORTED_LOCALES:
            raise UnsupportedLocale(locale)
```
Also locales is static and delivered with mimesis package. Only contributors can add or change locales with pull request (all tests should pass).

Heres docs for locale contributors:
https://github.com/lk-geimfari/mimesis/tree/master/mimesis/data/locale_template

imho, adding file checking is unnecessary

Reply:
Absolutely agree with @duckyou.